src/lapack/gpu/lacpy.cu

```diff
@@ -105,6 +105,80 @@ __global__ void lacpy(cublasFillMode_t uplo, const unsigned m, const unsigned n,
 }
 }
 
+static whip::memcpy_kind get_lacpy_memcpy_kind(const void* src, const void* dst) {
+// If HIP is version 5.6 or newer, avoid the use of hipMemcpyDefault as it is
+// buggy with 2D memcpy.  Instead try to infer the memory type using
+// hipPointerGetAttributes. See:
+// - https://github.com/ROCm/clr/commit/56daa6c4891b43ec233e9c63f755e3f7b45842b4
+// - https://github.com/ROCm/clr/commit/d3bfb55d7a934355257a72fab538a0a634b43cad
+//
+// Note that hipMemoryTypeManaged is not available in older versions, but it is
+// already available in 5.3 so we don't do a separate check for availability.
+#if defined(DLAF_WITH_HIP) && HIP_VERSION >= 50600000
+  constexpr auto get_memory_type = [](const void* p) {
+    hipPointerAttribute_t attributes{};
+    hipError_t status = hipPointerGetAttributes(&attributes, p);
+    if (status == hipErrorInvalidValue) {
+      // If HIP returns hipErrorInvalidValue we assume it's due
+      // to HIP not recognizing a non-HIP allocated pointer as
+      // host memory, and we assume the type is host.
+      return hipMemoryTypeHost;
+    }
+    else if (status != hipSuccess) {
+      throw whip::exception(status);
+    }
+
+    switch (attributes.type) {
+#if HIP_VERSION >= 60000000
+      case hipMemoryTypeUnregistered:
+        [[fallthrough]];
+#endif
+      case hipMemoryTypeHost:
+        return hipMemoryTypeHost;
+      case hipMemoryTypeArray:
+        [[fallthrough]];
+      case hipMemoryTypeDevice:
+        return hipMemoryTypeDevice;
+      case hipMemoryTypeUnified:
+        return hipMemoryTypeUnified;
+      case hipMemoryTypeManaged:
+        return hipMemoryTypeManaged;
+      default:
+        DLAF_UNREACHABLE_PLAIN;
+    }
+  };
+
+  hipMemoryType src_type = get_memory_type(src);
+  hipMemoryType dst_type = get_memory_type(dst);
+
+  if (src_type == hipMemoryTypeDevice && dst_type == hipMemoryTypeHost) {
+    return whip::memcpy_device_to_host;
+  }
+  else if (src_type == hipMemoryTypeHost && dst_type == hipMemoryTypeDevice) {
+    return whip::memcpy_host_to_device;
+  }
+  else if (src_type == hipMemoryTypeDevice && dst_type == hipMemoryTypeDevice) {
+    return whip::memcpy_device_to_device;
+  }
+  else if (src_type == hipMemoryTypeManaged || src_type == hipMemoryTypeUnified ||
+           dst_type == hipMemoryTypeManaged || dst_type == hipMemoryTypeUnified) {
+    return whip::memcpy_default;
+  }
+  else if (src_type == hipMemoryTypeHost && dst_type == hipMemoryTypeHost) {
+    DLAF_ASSERT(
+        false,
+        "Attempting to do a HIP lacpy with host source and destination, use the CPU lacpy instead");
+  }
+  else {
+    DLAF_ASSERT(false,
+                "Attempting to do a HIP lacpy with unsupported source and destination memory type",
+                src_type, dst_type);
+  }
```

Comment:
I've slightly refactored this. It's still a bit of a mess but perhaps passable. I now:
- Map a `hipErrorInvalidValue` from `hipPointerGetAttributes` to `hipMemoryTypeHost`. From a small test it seems like a `malloc`'d pointer does not go through the same path as a `hipHostMalloc`'d pointer, but we still need to cover the case of `hipMemoryTypeHost`, so we can treat both the same.
- Map `hipMemoryTypeUnregistered` (if available) to `hipMemoryTypeHost`.
- Map `hipMemoryTypeArray` to `hipMemoryTypeDevice` because the HIP documentation states that it's always on a device.
- If any pointer is `hipMemoryTypeManaged` or `hipMemoryTypeUnified` the memcpy kind is `hipMemcpyDefault` because I don't know what else to do with them. This may be too conservative, but at least I'm not making any assumptions that I'm not sure about. This case was previously also using `hipMemcpyDefault` so it should not be worse than before.

All of the above is only done if HIP is 5.6 or newer. I'm hoping that we can put an upper bound on the workaround maybe with 6.1 or whichever version properly fixes the behaviour.